File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/inventory/items/ItemRecyclerFragment.java
package com.habitrpg.android.habitica.ui.fragments.inventory.items;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.InventoryRepository;
import com.habitrpg.android.habitica.events.commands.OpenMenuItemCommand;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.models.inventory.Egg;
import com.habitrpg.android.habitica.models.inventory.Food;
import com.habitrpg.android.habitica.models.inventory.HatchingPotion;
import com.habitrpg.android.habitica.models.inventory.Item;
import com.habitrpg.android.habitica.models.inventory.Pet;
import com.habitrpg.android.habitica.models.inventory.QuestContent;
import com.habitrpg.android.habitica.models.inventory.SpecialItem;
import com.habitrpg.android.habitica.models.user.User;
import com.habitrpg.android.habitica.ui.adapter.inventory.ItemRecyclerAdapter;
import com.habitrpg.android.habitica.ui.fragments.BaseFragment;
import com.habitrpg.android.habitica.ui.helpers.RecyclerViewEmptySupport;
import com.habitrpg.android.habitica.ui.helpers.SafeDefaultItemAnimator;
import com.habitrpg.android.habitica.ui.menu.MainDrawerBuilder;

import org.greenrobot.eventbus.EventBus;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.realm.OrderedRealmCollection;

public class ItemRecyclerFragment extends BaseFragment {

    @Inject
    InventoryRepository inventoryRepository;

    private static final String ITEM_TYPE_KEY = "CLASS_TYPE_KEY";
    @BindView(R.id.recyclerView)
    public RecyclerViewEmptySupport recyclerView;
    @BindView(R.id.emptyView)
    public View emptyView;
    @BindView(R.id.empty_text_view)
    public TextView emptyTextView;
    @BindView(R.id.titleTextView)
    public TextView titleView;
    @BindView(R.id.footerTextView)
    public TextView footerView;
    @BindView(R.id.openMarketButton)
    public Button openMarketButton;
    public ItemRecyclerAdapter adapter;
    public String itemType;
    public String itemTypeText;
    public boolean isHatching;
    public boolean isFeeding;
    public Item hatchingItem;
    public Pet feedingPet;
    @Nullable
    public User user;
    LinearLayoutManager layoutManager = null;

    private View view;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        if (view == null) {
            view = inflater.inflate(R.layout.fragment_items, container, false);
        }
        unbinder = ButterKnife.bind(this, view);

        recyclerView.setEmptyView(emptyView);
        emptyTextView.setText(getString(R.string.empty_items, itemTypeText));

        android.support.v4.app.FragmentActivity context = getActivity();

        layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();

        if (layoutManager == null) {
            layoutManager = new LinearLayoutManager(context);

            recyclerView.setLayoutManager(layoutManager);
        }

        adapter = (ItemRecyclerAdapter) recyclerView.getAdapter();
        if (adapter == null) {
            adapter = new ItemRecyclerAdapter(null, true);
            adapter.context = this.getActivity();
            adapter.isHatching = this.isHatching;
            adapter.isFeeding = this.isFeeding;
            adapter.fragment = this;
            if (this.hatchingItem != null) {
                adapter.hatchingItem = this.hatchingItem;
            }
            if (this.feedingPet != null) {
                adapter.feedingPet = this.feedingPet;
            }
            recyclerView.setAdapter(adapter);

            compositeSubscription.add(adapter.getSellItemEvents()
                    .flatMap(item -> inventoryRepository.sellItem(user, item))
                    .subscribe(item -> {}, RxErrorHandler.handleEmptyError()));

            compositeSubscription.add(adapter.getQuestInvitationEvents()
                    .flatMap(quest -> inventoryRepository.inviteToQuest(quest))
                            .subscribe(group -> {
                                OpenMenuItemCommand event1 = new OpenMenuItemCommand();
                                event1.identifier = MainDrawerBuilder.INSTANCE.getSIDEBAR_PARTY();
                                EventBus.getDefault().post(event1);
                            }, RxErrorHandler.handleEmptyError()));
        }
        recyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL));
        recyclerView.setItemAnimator(new SafeDefaultItemAnimator());

        if (savedInstanceState != null) {
            this.itemType = savedInstanceState.getString(ITEM_TYPE_KEY, "");
        }

        if (this.isHatching) {
            getDialog().requestWindowFeature(Window.FEATURE_NO_TITLE);
            this.titleView.setText(getString(R.string.hatch_with, this.hatchingItem.getText()));
            this.titleView.setVisibility(View.VISIBLE);
            this.footerView.setText(getString(R.string.hatching_market_info));
            this.footerView.setVisibility(View.VISIBLE);
            this.openMarketButton.setVisibility(View.VISIBLE);
        } else if (this.isFeeding) {
            getDialog().requestWindowFeature(Window.FEATURE_NO_TITLE);
            this.titleView.setText(getString(R.string.dialog_feeding, this.feedingPet.getColorText(), this.feedingPet.getAnimalText()));
            this.titleView.setVisibility(View.VISIBLE);
            this.footerView.setText(getString(R.string.feeding_market_info));
            this.footerView.setVisibility(View.VISIBLE);
            this.openMarketButton.setVisibility(View.VISIBLE);
        } else {
            this.titleView.setVisibility(View.GONE);
            this.footerView.setVisibility(View.GONE);
            this.openMarketButton.setVisibility(View.GONE);
        }

        return view;
    }

    @Override
    public void onDestroy() {
        inventoryRepository.close();
        super.onDestroy();
    }

    @Override
    public void injectFragment(AppComponent component) {
        component.inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        this.loadItems();
    }

    @Override
    public void onResume() {
        if ((this.isHatching || this.isFeeding) && getDialog().getWindow() != null) {
            android.view.WindowManager.LayoutParams params = getDialog().getWindow().getAttributes();
            params.width = ViewGroup.LayoutParams.MATCH_PARENT;
            params.verticalMargin = 60;
            getDialog().getWindow().setAttributes(params);
        }

        super.onResume();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(ITEM_TYPE_KEY, this.itemType);
    }

    private void loadItems() {
        Class<? extends Item> itemClass = null;
        switch (itemType) {
            case "eggs":
                itemClass = Egg.class;
                break;
            case "hatchingPotions":
                itemClass = HatchingPotion.class;
                break;
            case "food":
                itemClass = Food.class;
                break;
            case "quests":
                itemClass = QuestContent.class;
                break;
            case "special":
                itemClass = SpecialItem.class;
                break;
        }
        inventoryRepository.getOwnedItems(itemClass, user).first().subscribe(items -> {
            if (items.size() > 0) {
                adapter.updateData((OrderedRealmCollection<Item>) items);
            }
        }, RxErrorHandler.handleEmptyError());

        compositeSubscription.add(inventoryRepository.getOwnedPets().subscribe(adapter::setOwnedPets, RxErrorHandler.handleEmptyError()));
    }

    @OnClick(R.id.openMarketButton)
    public void onOpenMarketClicked() {
        dismiss();
        openMarket();
    }

    @OnClick(R.id.openEmptyMarketButton)
    public void onEmptyOpenMarketButtonClicked() {
        openMarket();
    }

    private void openMarket() {
        OpenMenuItemCommand command = new OpenMenuItemCommand();
        command.identifier = MainDrawerBuilder.INSTANCE.getSIDEBAR_SHOPS();
        EventBus.getDefault().post(command);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/inventory/items/ItemRecyclerFragment.kt
package com.habitrpg.android.habitica.ui.fragments.inventory.items

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.databinding.FragmentItemsBinding
import com.habitrpg.android.habitica.extensions.addCancelButton
import com.habitrpg.android.habitica.extensions.addCloseButton
import com.habitrpg.android.habitica.models.Skill
import com.habitrpg.android.habitica.models.inventory.Egg
import com.habitrpg.android.habitica.models.inventory.HatchingPotion
import com.habitrpg.android.habitica.models.inventory.Item
import com.habitrpg.android.habitica.models.inventory.SpecialItem
import com.habitrpg.android.habitica.models.user.OwnedItem
import com.habitrpg.android.habitica.models.user.User
import com.habitrpg.android.habitica.ui.activities.BaseActivity
import com.habitrpg.android.habitica.ui.activities.MainActivity
import com.habitrpg.android.habitica.ui.activities.SkillMemberActivity
import com.habitrpg.android.habitica.ui.adapter.inventory.ItemRecyclerAdapter
import com.habitrpg.android.habitica.ui.fragments.BaseFragment
import com.habitrpg.android.habitica.ui.fragments.skills.SkillDialogBottomSheetFragment
import com.habitrpg.android.habitica.ui.helpers.SafeDefaultItemAnimator
import com.habitrpg.android.habitica.ui.viewmodels.MainUserViewModel
import com.habitrpg.android.habitica.ui.viewmodels.inventory.items.ItemListViewModel
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaAlertDialog
import com.habitrpg.android.habitica.ui.views.dialogs.OpenedMysteryitemDialog
import com.habitrpg.common.habitica.extensions.loadImage
import com.habitrpg.common.habitica.extensions.observeOnce
import com.habitrpg.common.habitica.helpers.EmptyItem
import com.habitrpg.common.habitica.helpers.MainNavigationController
import com.habitrpg.common.habitica.helpers.launchCatching
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class ItemRecyclerFragment :
    BaseFragment<FragmentItemsBinding>(),
    SwipeRefreshLayout.OnRefreshListener {
    val viewModel: ItemListViewModel by viewModels()

    @Inject
    lateinit var userViewModel: MainUserViewModel

    var user: User? = null
    var adapter: ItemRecyclerAdapter? = null
    var itemType: String?
        get() = arguments?.getString(ARG_ITEM_TYPE)
        set(value) {
            val args = arguments ?: Bundle().also { arguments = it }
            args.putString(ARG_ITEM_TYPE, value)
        }
    var itemTypeText: String?
        get() = arguments?.getString(ARG_ITEM_TYPE_TEXT)
        set(value) {
            val args = arguments ?: Bundle().also { arguments = it }
            args.putString(ARG_ITEM_TYPE_TEXT, value)
        }
    private var selectedSpecialItem: SpecialItem? = null
    private var specialSkills: MutableList<Skill> = mutableListOf()
    internal var layoutManager: androidx.recyclerview.widget.LinearLayoutManager? = null

    override var binding: FragmentItemsBinding? = null

    override fun createBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
    ): FragmentItemsBinding = FragmentItemsBinding.inflate(inflater, container, false)

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        getSpecialSkills()

        binding?.refreshLayout?.setOnRefreshListener(this)
        val buttonMethod = {
            if (itemType == "quests") {
                MainNavigationController.navigate(R.id.questShopFragment)
            } else {
                openMarket()
            }
        }
        binding?.recyclerView?.emptyItem =
            EmptyItem(
                getString(R.string.no_x, itemTypeText ?: itemType),
                when (itemType) {
                    "food" -> getString(R.string.empty_food_description)
                    "quests" -> getString(R.string.empty_quests_description)
                    "special" -> getString(R.string.empty_special_description_subscribed)
                    "eggs" -> getString(R.string.empty_eggs_description)
                    "hatchingPotions" -> getString(R.string.empty_potions_description)
                    else -> ""
                },
                when (itemType) {
                    "eggs" -> R.drawable.icon_eggs
                    "hatchingPotions" -> R.drawable.icon_hatchingpotions
                    "food" -> R.drawable.icon_food
                    "quests" -> R.drawable.icon_quests
                    "special" -> R.drawable.icon_special
                    else -> null
                },
                false,
                if (itemType == "special") null else buttonMethod,
            )

        layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        binding?.recyclerView?.layoutManager = layoutManager
        binding?.recyclerView?.itemAnimator = SafeDefaultItemAnimator()

        userViewModel.user.observeOnce(this) {
            if (it != null) {
                user = it
                adapter?.user = it
            }
        }

        binding?.titleTextView?.visibility = View.GONE
        setAdapter()
        this.loadItems()
    }

    private fun getSpecialSkills() {
        // Get special skills for description of special items
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.specialSkills.collect { skills -> specialSkills = skills }
        }
    }

    private fun setAdapter() {
        val context = activity

        adapter = binding?.recyclerView?.adapter as? ItemRecyclerAdapter
        if (adapter == null) {
            context?.let {
                adapter = ItemRecyclerAdapter(context)
            }
            binding?.recyclerView?.adapter = adapter
        }
        adapter?.onUseSpecialItem = { specialItem ->
            val specialSkill = specialSkills.find { it.key == specialItem.key }
            if (specialSkill != null) {
                val skillIdentifier = "shop_"
                val bottomSheet =
                    SkillDialogBottomSheetFragment.newInstance(
                        skillTitle = specialSkill.text,
                        skillDescription = specialSkill.notes ?: "",
                        skillKey = specialSkill.key,
                        skillPath = skillIdentifier,
                        isTransformationItem = true,
                        onUseSkill = {
                            onSpecialItemSelected(specialItem)
                        },
                    )
                bottomSheet.show(childFragmentManager, "SkillDialogBottomSheet")
            }
        }
        adapter?.onSellItem = { item, ownedItem ->
            showSellItemConfirmation(item, ownedItem)
        }
        adapter?.onQuestInvitation = {
            viewModel.inviteToQuest(it)
        }
        adapter?.onOpenMysteryItem = {
            viewLifecycleOwner.lifecycleScope.launchCatching {
                val item = viewModel.openMysteryItem(user) ?: return@launchCatching
                val activity = activity as? MainActivity
                if (activity != null) {
                    val dialog = OpenedMysteryitemDialog(activity)
                    dialog.isCelebratory = true
                    dialog.setTitle(R.string.mystery_item_title)
                    dialog.binding.iconView.loadImage("shop_${item.key}")
                    dialog.binding.titleView.text = item.text
                    dialog.binding.descriptionView.text = item.notes
                    dialog.addButton(R.string.equip, true) { _, _ ->
                        item.key?.let { mysteryItem ->
                            viewModel.equip(
                                "equipped",
                                mysteryItem,
                            )
                        }
                    }
                    dialog.addCloseButton()
                    dialog.enqueue()
                }
            }
        }
        adapter?.onStartHatching = { showHatchingDialog(it) }
        adapter?.onHatchPet = { pet, egg -> hatchPet(pet, egg) }
        adapter?.onCreateNewParty = { createNewParty() }
        adapter?.itemType = itemType ?: ""
        adapter?.itemText =
            (if (itemType == "hatchingPotions") context?.getString(R.string.potions) else itemTypeText)
                ?: ""
        adapter?.onOpenShop = {
            if (itemType == "quests") {
                MainNavigationController.navigate(R.id.questShopFragment)
            } else {
                openMarket()
            }
        }
    }

    private fun showHatchingDialog(item: Item) {
        val fragment = ItemDialogFragment()
        if (item is Egg) {
            fragment.itemType = "hatchingPotions"
            fragment.hatchingItem = item
        } else {
            fragment.itemType = "eggs"
            fragment.hatchingItem = item
        }
        fragment.isHatching = true
        fragment.isFeeding = false
        parentFragmentManager.let { fragment.show(it, "hatchingDialog") }
    }

    override fun onRefresh() {
        binding?.refreshLayout?.isRefreshing = true
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.refreshUser()
            binding?.refreshLayout?.isRefreshing = false
        }
    }

    private fun hatchPet(
        potion: HatchingPotion,
        egg: Egg,
    ) {
        (activity as? BaseActivity)?.let {
            viewModel.hatchPet(potion, egg, it)
        }
    }

    private fun createNewParty() {
        val alert = context?.let { HabiticaAlertDialog(it) }
        alert?.setTitle(R.string.quest_party_required_title)
        alert?.setMessage(R.string.quest_party_required_description)
        alert?.addButton(R.string.create_new_party, true, false) { _, _ ->
            viewLifecycleOwner.lifecycleScope.launchCatching {
                user?.let {
                    viewModel.createParty(
                        it,
                        getString(R.string.usernames_party, user?.profile?.name),
                    )
                }
            }
        }
        alert?.addButton(R.string.close, false) { _, _ ->
            alert.dismiss()
        }
        alert?.show()
    }

    private fun loadItems() {
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.ownedItems.collect { adapter?.data = it }
        }
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.items.collect {
                adapter?.items = it
            }
        }
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.pets.collect { adapter?.setExistingPets(it) }
        }
        viewLifecycleOwner.lifecycleScope.launchCatching {
            viewModel.ownedPets.collect { adapter?.setOwnedPets(it) }
        }
    }

    private fun openMarket() {
        MainNavigationController.navigate(R.id.marketFragment)
    }

    private fun onSpecialItemSelected(specialItem: SpecialItem) {
        selectedSpecialItem = specialItem
        val intent = Intent(activity, SkillMemberActivity::class.java)
        memberSelectionResult.launch(intent)
    }

    private val memberSelectionResult =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == Activity.RESULT_OK) {
                useSpecialItem(selectedSpecialItem, it.data?.getStringExtra("member_id"))
            }
        }

    private fun useSpecialItem(
        specialItem: SpecialItem?,
        memberID: String? = null,
    ) {
        if (specialItem == null || memberID == null) {
            return
        }
        lifecycleScope.launchCatching {
            viewModel.useSkill(specialItem.key, specialItem.target, memberID)
            displaySpecialItemResult(specialItem)
        }
    }

    private fun displaySpecialItemResult(specialItem: SpecialItem?) {
        if (!isAdded) return

        val activity = activity as? MainActivity
        activity?.let {
            HabiticaSnackbar.showSnackbar(
                it.snackbarContainer,
                context?.getString(R.string.used_skill_without_mana, specialItem?.text),
                HabiticaSnackbar.SnackbarDisplayType.BLUE,
            )
        }

        loadItems()
    }

    private fun showSellItemConfirmation(
        item: Item,
        ownedItem: OwnedItem,
    ) {
        val dialog = HabiticaAlertDialog(requireContext())
        dialog.setTitle(getString(R.string.sell_confirmation_title, item.text))
        dialog.addButton(
            getString(R.string.sell, item.value),
            isPrimary = true,
            isDestructive = true,
        ) { _, _ ->
            lifecycleScope.launchCatching {
                viewModel.sellItem(ownedItem)
            }
        }
        dialog.addCancelButton()
        dialog.show()
    }

    companion object {
        const val ARG_ITEM_TYPE = "CLASS_TYPE_KEY"
        const val ARG_ITEM_TYPE_TEXT = "CLASS_TYPE_TEXT_KEY"
    }
}
